Matrix.cpp

```diff
@@ -347,10 +277,56 @@ ofstream& operator<<(ofstream& fileOut, const Matrix& mat) {
 // @param int c - the number of columns
 // @param double value - set each element in the matrix to this value
 void Matrix::setFields(int r, int c, double value) {
-	// write a check here for the parameters. e.g. r >= 0
 	nRows = r;
 	nCols = c;
 	data = vector<vector<double>>(r, vector<double>(c, value));
-	// setLimitsToData();
+}
+
+// Set the class fields
+// @param const vector<vector<double>>& d - the input vector from which to copy
+void Matrix::setFields(const vector<vector<double>>& d) {
+	data = d;
+	nRows = data.size();
+	nCols = data[0].size();
+}
+
+// Set the class fields
+// @param const Matrix& mat - the input Matrix from which to copy
+void Matrix::setFields(const Matrix& mat) {
+	data = mat.data;
+	nRows = data.size();
+	nCols = data[0].size();
+}
+
+// Set the class fields
+// @param const SubMatrix& sm - the input SubMattrix from which to copy
+void Matrix::setFields(const SubMatrix& sm) {
+	nRows = sm.nRows;
+	nCols = sm.nCols;
+	data = vector<vector<double>>(sm.nRows, vector<double>(sm.nCols));
+	for (int i = 0; i < nRows; ++i) {
+		for (int j = 0; j < nCols; ++j) {
+			data[i][j] = sm.data->operator()(i + sm.top, j + sm.left);

```

Comment:
Overload the () operator in SubMatrix to automatically offset the index by its top and left fields.